### PR TITLE
Move InstanceManager to pkg/prom/instance

### PR DESCRIPTION
**Pull request by rfratto**
This commit migrates `InstanceManager` to `instance.BasicManager`, and adds a `instance.Manager` interface to represent it. This is to reduce duplication as we add more implementations of the `instance.Manager` interface (including removing `ha.InstanceManager`, which was a duplicate of that interface).

While creating this commit, I noticed that there was a leaked locked mutex when `DeleteConfig` fails. That has been fixed with this commit.

Related to the work for #141.

### CODE REVIEW

**Comment by rfratto** [comment]
Marking ready for review now that #145 is merged.

File: pkg/prom/instance/manager.go [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+	// DefaultBasicManagerConfig is the default config for the BasicManager.
+	DefaultBasicManagerConfig = BasicManagerConfig{
+		InstanceRestartBackoff: 5 * time.Second,
+	}
+)
+
+// Manager represents a set of methods for manipulating running instances at
+// runtime.
+type Manager interface {
+	// ListInstances should return all currently managed instances running
```
Comment (gotjosh) [inline pkg/prom/instance/manager.go:37]:
```suggestion
	// ListInstances returns all currently managed instances running
```

File: pkg/prom/instance/manager.go [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+	// within the Manager. The key will be the instance name from their config.
+	ListInstances() map[string]ManagedInstance
+
+	// ListConfigs will return the config objects associated with a managed
```
Comment (gotjosh) [inline pkg/prom/instance/manager.go:41]:
```suggestion
	// ListConfigs returns the config objects associated with a managed
```

File: pkg/prom/instance/manager.go [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+	// instance. The key will be the Name field from Config.
+	ListConfigs() map[string]Config
+
+	// ApplyConfig should create a new Config or update an existing Config if
```
Comment (gotjosh) [inline pkg/prom/instance/manager.go:45]:
```suggestion
	// ApplyConfig creates a new Config or updates an existing Config if
```

File: pkg/prom/instance/manager.go [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+	// one with Config.Name already exists.
+	ApplyConfig(Config) error
+
+	// DeleteConfig should delete a given managed instance from its Config.Name.
```
Comment (gotjosh) [inline pkg/prom/instance/manager.go:49]:
```suggestion
	// DeleteConfig deletes a given managed instance based on its Config.Name.
```

File: pkg/prom/instance/manager.go [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+	// Stop on a process, you will deadlock.
+	mut       sync.Mutex
+	processes map[string]*managedProcess
+
+	launch   Factory
+	validate ConfigValidator
+}
+
+// managedProcess represents a goroutine running a ManagedInstance. cancel
+// requests that the goroutine should should down. done will be closed after the
```
Comment (gotjosh) [inline pkg/prom/instance/manager.go:87]:
```suggestion
// requests that the goroutine should shutdown. done will be closed after the
```

**Comment by rfratto** [comment]
> I take it this is a no-op (minus the fix for the DeleteConfig) then?
Right, no (intended) behaviors will change with this.

**Pull request merged**